controllers/secret_controller.go

```diff
@@ -0,0 +1,370 @@
+				err := fmt.Errorf("target secret is immutable")
+				logger.Error(err, "Failed to modify target secret, secret is set to immutable", "Secret", fetchedSecret.Name, "TargetContext", target)
+				break TargetSecrets
+			}
+
+			if requiresUpdate(sec, fetchedSecret) {
+				logger.Info("Modifying secret in target cluster", "Secret", sec.Name, "TargetContext", target)
+				syncSecrets(sec, fetchedSecret)
+				if err = remoteClient.Update(ctx, fetchedSecret); err != nil {
+					logger.Error(err, "Failed to sync target secret for matching payloads", "Secret", fetchedSecret.Name, "TargetContext", target)
+					break TargetSecrets
+				}
+			}
+		}
+		if err != nil {
+			cond.Status = corev1.ConditionFalse
+		} else {
+			cond.Status = corev1.ConditionTrue
+		}
+
+		timeNow := metav1.Now()
+		cond.LastTransitionTime = &timeNow
+		rsec.Status.Conditions = append(rsec.Status.Conditions, cond)
+	}
+
+	// Update the ReplicatedSecret's Status
+	err = localClient.Status().Patch(ctx, rsec, patch)
+	if err != nil {
+		logger.Error(err, "Failed to update replicated secret last transition time", "ReplicatedSecret", req.NamespacedName)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// If any cluster had failed state, retry
+	for _, cond := range rsec.Status.Conditions {
+		if cond.Status == corev1.ConditionFalse {
+			return ctrl.Result{Requeue: true}, fmt.Errorf("replication failed")
+		}
+	}
+	return ctrl.Result{}, err
+}
+
+func requiresUpdate(source, dest client.Object) bool {
+	if source.GetGeneration() != 0 {
+		// Compare using generation
+		return source.GetGeneration() > dest.GetGeneration()
+	} else {
+		if srcHash, found := source.GetAnnotations()[api.ResourceHashAnnotation]; found {
+			destHash := dest.GetAnnotations()[api.ResourceHashAnnotation]
+			return srcHash != destHash
+		}
+		// Verify data wasn't manually updated without hash changes..
+		if sourceSec, valid := source.(*corev1.Secret); valid {
+			if destSec, valid := dest.(*corev1.Secret); valid {
+				return !reflect.DeepEqual(sourceSec.Data, destSec.Data)
+			}
+		}
+	}
+	return false
+}
+
+func syncSecrets(src, dest *corev1.Secret) {
+	origMeta := dest.ObjectMeta
+	src.DeepCopyInto(dest)
+	dest.ObjectMeta = origMeta
+	dest.OwnerReferences = []metav1.OwnerReference{}
+
+	// sync annotations, src is more important
+	if dest.GetAnnotations() == nil {
+		dest.Annotations = make(map[string]string)
+	}
+
+	for k, v := range src.Annotations {
+		dest.Annotations[k] = v
+	}
+
+	// sync labels, src is more important
+	if dest.GetLabels() == nil {
+		dest.Labels = make(map[string]string)
+	}
+
+	for k, v := range src.Labels {
+		dest.Labels[k] = v
+	}
+}
+
+func (s *SecretSyncController) verifyHashAnnotation(ctx context.Context, sec *corev1.Secret) error {
+	hash := utils.DeepHashString(sec.Data)
+	if sec.GetAnnotations() == nil {
+		sec.Annotations = make(map[string]string)
+	}
+	if existingHash, found := sec.GetAnnotations()[api.ResourceHashAnnotation]; !found || (existingHash != hash) {
```

Comment:
If the hash has changed, does that mean we need to update to match the source secret?